fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/OCFLPersistentStorageSession.java

```diff
@@ -166,7 +279,48 @@ public void rollback() throws PersistentStorageException {
             // No changes to rollback
             return;
         }
-        // rollback changes
+
+        if (!state.equals(State.COMMIT_FAILED) && !state.equals(State.PREPARE_FAILRED)) {
+            throw new PersistentStorageException("This session cannot be rolled back in this state: " + state);
+        }
+
+        this.state = State.ROLLING_BACK;
```

Comment:
Hmm, I wonder if this should wait for all the `persist` calls to finish. That'll mainly only matter if this method is going to handle user requested transaction cancels. For post-commit failure situations, it probably isn't necessary since the commit will have already blocked to wait for them, so it'd be a programming error to reach this method without passing through commit first.

Reply:
Yes. I think this makes sense.  Good catch.